### ISSUE TITLE
Using keras model with torch.func.jacrev gives zero gradients 

### ISSUE THREAD

**Issue by francescoPolizzi** [keras-team/keras#20279]
## Issue description

Trying computing the model jacobian using a keras model returns zeros
## Code example

```python
import keras
import torch
model = keras.Sequential(
    [keras.layers.Dense(1,activation='linear')]
)
model.build(input_shape=(1,1))
x = torch.tensor([1.0])
params = dict(model.named_parameters())
def f(params, x):
    prediction = torch.func.functional_call(
        model, params, x, strict=True,
    )
    print("prediction", prediction)
    print("prediction has grad", prediction.requires_grad)
    return prediction

J = torch.func.jacrev(f)(
    params,
    x,
)
```
Gives as output
```
prediction GradTrackingTensor(lvl=1, value=
    tensor([[0.7901]], grad_fn=<AddBackward0>)
)
prediction has grad False
{'_functional.torch_params.sequential_12/dense_97/bias': tensor([[[0.]]]),
 '_functional.torch_params.sequential_12/dense_97/kernel': tensor([[[[0.]]]])}
```
I am new to pytorch, maybe the problem is related with the fact that the prediction `requires_grad `is false ? I don't understand why if  I call the model outside `torch.func.jacrev` the `requires_grad` attribute of the prediction becomes True _(code below)_
```python
model(x).requires_grad
```
Maybe there is some compatbility issue using with the Keras model? 

Using a pure pytorch code it seems to work correctly
```python
model = torch.nn.Sequential(
    nn.Linear(1,1)
)

x = torch.tensor([1.0])
params = dict(model.named_parameters())
def f(params, x):
    prediction = torch.func.functional_call(
        model, params, x, strict=True,
    )
    print("prediction", prediction)
    print("prediction has grad", prediction.requires_grad)
    return prediction

J = torch.func.jacrev(f)(
    params,
    x,
)
J
```

```
{'0.weight': tensor([[[1.]]]), '0.bias': tensor([[1.]])}
```
## System Info
Python 3.9.5
```
keras                     3.5.0
torch                     2.4.0
torchaudio                2.4.0
torchmetrics              1.4.1
torchvision               0.19.0
```

**Comment by VarunS1997** [comment]
Have you taken a look at "Model.stateless_call()`? 

Theoretically, PyTorch should work directly but we have not particularly tested the functional_call(). Its functionality would depend on how the torch function is written.

**Comment by smartArancina** [comment]
Hi @VarunS1997 thanks for the answer.
Yes, I tried as you suggested and it seems to work on this dummy example
```python
import keras
import torch
model = keras.Sequential(
    [keras.layers.Dense(1,activation='linear')]
)
model.build(input_shape=(1,1))
x = torch.tensor([1.0])
params = [p.value for p in model.trainable_weights]
def f(params, x):
    # pytorch functional_call
    # prediction = torch.func.functional_call(
    #     model, params, x, strict=True,
    # )
    prediction = model.stateless_call(params, [], x)[0]
    print("prediction", prediction)
    print("prediction has grad", prediction.requires_grad)
    return prediction
#
J = torch.func.jacrev(f)(
    params,
    x,
)
J
```
```
prediction GradTrackingTensor(lvl=1, value=
    tensor([[-0.5455]], grad_fn=<AddBackward0>)
)
prediction has grad True
[tensor([[[[1.]]]]), tensor([[[1.]]])]
```
The problem is that I am using another pytorch package where the model differentations are done using `torch.func.functional_call` and I would not modify their source code to be compliant with keras. For this reason I ask you :
1) is there a workaround to make keras + Pytorch fully trasparent to pure pytorch code (`torch.func.functional_call`)? For exampIe _unwrapping_ the pytorch model from keras, make the computation with ext. libraries and then _re-wrapping_ trasparently will be a workaroud. I am asking this because I am moving from keras + Tensorflow to Keras + Pytorch (because I need some pyotorch external libraries) and the trasparancy above  will be a crucial aspect.
2) Just to understand better / curiosity: why the the keras model called with `torch.func.functional_call` has no gradient _prediction has grad False_ ? Is this the reason for which it doesn't work ? 
Thanks!

**Comment by fchollet** [comment]
> is there a workaround to make keras + Pytorch fully trasparent to pure pytorch code 

Keras is written in pure Python. When you are calling a Keras model with the PyTorch backend, you're just calling a Python function that contains torch ops. There is no magic at all. In fact Keras models inherit from the PyTorch `Module`.

We don't know why `functional_call` doesn't work here, but chances are the reasons are similar to why PyTorch `compile` usually doesn't work: it isn't Keras-specific, it just doesn't work for most Python functions.

**Comment by smartArancina** [comment]
Yes, I saw that the Keras inherit directly from Pytorch `Module`. I was just wondering if what is built around to pytorch `Module` with keras makes the difference between the pure pytorch working code and the not (keras). This just becouse I am planning to divulgate keras as main framework in bigger projects, so trasparency across ext Pytorch / Tf / Jax libraries is a key point (if this hypotesis agrees with keras ones of course)
Thanks !

**Comment by maitry63** [comment]
Hi @smartArancina , as discussed, this behavior is completely normal and expected. It happens because `Keras 3` and PyTorch manage variables differently under the hood.

Even though Keras 3 models can sit inside PyTorch code, Keras still runs calculations using its own internal tracking engine.

When you run` torch.func.functional_call`, PyTorch tries to replace the model's weights. However, Keras layers ignore these replaced weights and look at their own internal variables instead. Because functional_call cannot change those internal Keras variables, the math graph treats your weights as unchanging constants. This is why the gradients come out as all zeros.

Since the code is working exactly as designed for Keras 3 cross-framework compatibility, we are closing this issue! Feel free to open a new one if you run into any other problems.
